jmclient/jmclient/wallet_service.py

```diff
@@ -347,23 +360,31 @@ def transaction_monitor(self):
                 # is absurdly fast, this is considered acceptable compared with
                 # additional complexity.
                 self.log_new_tx(removed_utxos, added_utxos, txid)
-                self.processed_txids.append(txid)
+                self.processed_txids.add(txid)
 
             # first fire 'all' type callbacks, irrespective of if the
             # transaction pertains to anything known (but must
             # have correct label per above); filter on this Joinmarket wallet label,
             # or the external monitoring label:
-            if (self.bci.is_address_labeled(tx, self.get_wallet_name()) or
-                self.bci.is_address_labeled(tx, self.EXTERNAL_WALLET_LABEL)):
+            if (any(self.bci.is_address_labeled(addr, wallet_name) or
+                    self.bci.is_address_labeled(addr,
+                        self.EXTERNAL_WALLET_LABEL)
+                    for addr in res["details"])):
                 for f in self.callbacks["all"]:
                     # note we need no return value as we will never
                     # remove these from the list
                     f(txd, txid)
 
-            # The tuple given as the second possible key for the dict
-            # is such because txid is not always available
-            # at the time of callback registration).
-            possible_keys = [txid, tuple((x.scriptPubKey, x.nValue) for x in txd.vout)]
+            # txid is not always available at the time of callback registration.
+            # Migrate any callbacks registered under the provisional key, and
+            # leave a txid breadcrumb so check_callback_called can find it.
+            txos = tuple((x.scriptPubKey, x.nValue) for x in txd.vout)
+            for cb_type in ["unconfirmed", "confirmed"]:
+                callbacks = self.callbacks[cb_type]
+                reg = callbacks.get(txos)
+                if isinstance(reg, list):
+                    callbacks.setdefault(txid, [])[:0] = reg
```

Comment:
Oh, I've never seen this syntax. So does this just prepend one list to another? Like, the opposite of `extend`?
And if so, I'm curious, why put it at the start rather than the end of the list?

Reply:
> So does this just prepend one list to another?

Yes. `a[:0] = b` replaces the slice of `a` from its beginning to just before its index 0 with the contents of `b`, or in other words it prepends `b` to `a`.

> why put it at the start rather than the end of the list?

My assumption was that callbacks registered under the provisional key would have been registered prior to callbacks registered under the TxID key. I wanted to have the callbacks invoked in roughly the order in which they were registered.

Reply:
Yes; either way it certainly isn't wrong. I admit I didn't think about ordering at all (or let's say, I didn't allow the logic to specify any ordering, which tbh probably makes sense given async). So, all good.